File: JavaCryptoService/src/main/java/com/alfabank/crypto/service/EncryptionService.java
package com.alfabank.crypto.service;

import com.alfabank.crypto.crypto.CmsDecryptionProvider;
import com.alfabank.crypto.crypto.CmsEncryptionProvider;
import com.alfabank.crypto.dto.request.DecryptRequest;
import com.alfabank.crypto.dto.request.EncryptRequest;
import com.alfabank.crypto.dto.response.DecryptResponse;
import com.alfabank.crypto.dto.response.EncryptResponse;
import com.alfabank.crypto.exception.CryptoOperationException;
import com.alfabank.crypto.keystore.KeystoreManager;
import com.alfabank.crypto.model.EncryptionDetail;
import com.alfabank.crypto.model.OperationType;
import com.alfabank.crypto.repository.EncryptionDetailRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.io.ByteArrayInputStream;
import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;
import java.security.PrivateKey;
import java.security.cert.CertificateExpiredException;
import java.security.cert.CertificateFactory;
import java.security.cert.CertificateNotYetValidException;
import java.security.cert.X509Certificate;
import java.time.Instant;
import java.util.Base64;

@Service
public class EncryptionService {

    private static final Logger log = LoggerFactory.getLogger(EncryptionService.class);

    private final CmsEncryptionProvider encryptionProvider;
    private final CmsDecryptionProvider decryptionProvider;
    private final KeystoreManager keystoreManager;
    private final AuditedOperation auditedOperation;
    private final EncryptionDetailRepository detailRepository;

    public EncryptionService(CmsEncryptionProvider encryptionProvider,
                             CmsDecryptionProvider decryptionProvider,
                             KeystoreManager keystoreManager,
                             AuditedOperation auditedOperation,
                             EncryptionDetailRepository detailRepository) {
        this.encryptionProvider = encryptionProvider;
        this.decryptionProvider = decryptionProvider;
        this.keystoreManager = keystoreManager;
        this.auditedOperation = auditedOperation;
        this.detailRepository = detailRepository;
    }

    @Transactional
    public EncryptResponse encrypt(EncryptRequest request) {
        log.info("Encrypt requested");
        return auditedOperation.run(OperationType.ENCRYPT, null, null, "ENCRYPT_FAILED",
                (msg, cause) -> new CryptoOperationException("Encrypt failed: " + msg, cause),
                op -> {
                    byte[] plaintext = Base64.getDecoder().decode(request.plaintext());
                    log.debug("[{}] plaintextSize={} bytes", op.getId(), plaintext.length);
                    X509Certificate recipientCert = parseCert(Base64.getDecoder().decode(request.recipientCertificate()));
                    log.debug("[{}] recipient: {}", op.getId(), recipientCert.getSubjectX500Principal().getName());
                    byte[] packed = packWithFilename(plaintext, request.filename());
                    byte[] ciphertext = encryptionProvider.encrypt(packed, recipientCert);
                    saveEncryptDetail(op.getId(), recipientCert.getSubjectX500Principal().getName(),
                            plaintext.length, ciphertext.length, request.filename());
                    log.info("[{}] encrypted: {}→{} bytes", op.getId(), plaintext.length, ciphertext.length);
                    EncryptResponse response = new EncryptResponse(op.getId(),
                            Base64.getEncoder().encodeToString(ciphertext), Instant.now());
                    return AuditedOperation.Outcome.of(response, ciphertext);
                });
    }

    @Transactional
    public DecryptResponse decrypt(DecryptRequest request) {
        log.info("Decrypt requested: alias={}", request.keyAlias());
        return auditedOperation.run(OperationType.DECRYPT, null, request.keyAlias(), "DECRYPT_FAILED",
                (msg, cause) -> new CryptoOperationException("Decrypt failed: " + msg, cause),
                op -> {
                    byte[] ciphertext = Base64.getDecoder().decode(request.ciphertext());
                    log.debug("[{}] ciphertextSize={} bytes", op.getId(), ciphertext.length);
                    PrivateKey privateKey = keystoreManager.getPrivateKey(request.keyAlias());
                    byte[] decrypted = decryptionProvider.decrypt(ciphertext, privateKey);
                    UnpackedContent unpacked = unpackFilename(decrypted);
                    log.info("[{}] decrypted: {}→{} bytes, filename={}", op.getId(), ciphertext.length,
                            unpacked.content().length, unpacked.filename());
                    DecryptResponse response = new DecryptResponse(op.getId(),
                            Base64.getEncoder().encodeToString(unpacked.content()), unpacked.filename(), Instant.now());
                    return AuditedOperation.Outcome.of(response, unpacked.content());
                });
    }

    private X509Certificate parseCert(byte[] derBytes) throws Exception {
        CertificateFactory cf = CertificateFactory.getInstance("X.509", "BC");
        X509Certificate cert = (X509Certificate) cf.generateCertificate(new ByteArrayInputStream(derBytes));
        try {
            cert.checkValidity();
        } catch (CertificateExpiredException | CertificateNotYetValidException e) {
            throw new CryptoOperationException("Recipient certificate is not valid: " + e.getMessage(), e);
        }
        return cert;
    }

    private void saveEncryptDetail(String opId, String recipientDn, int inputSize, int outputSize, String filename) {
        EncryptionDetail d = new EncryptionDetail();
        d.setOperationId(opId);
        d.setAlgorithm("AES256_GCM");
        d.setRecipientDn(recipientDn);
        d.setInputSize(inputSize);
        d.setOutputSize(outputSize);
        d.setOriginalFilename(filename);
        detailRepository.save(d);
    }

    private static final byte FORMAT_MARKER_PACKED_FILENAME = (byte) 0x01;

    private byte[] packWithFilename(byte[] content, String filename) {
        byte[] nameBytes = (filename != null && !filename.isBlank())
                ? filename.getBytes(StandardCharsets.UTF_8)
                : new byte[0];
        ByteBuffer buf = ByteBuffer.allocate(1 + 4 + nameBytes.length + content.length);
        buf.put(FORMAT_MARKER_PACKED_FILENAME);
        buf.putInt(nameBytes.length);
        buf.put(nameBytes);
        buf.put(content);
        return buf.array();
    }

    private UnpackedContent unpackFilename(byte[] packed) {
        if (packed.length == 0 || packed[0] != FORMAT_MARKER_PACKED_FILENAME) {
            return new UnpackedContent(null, packed);
        }
        if (packed.length < 1 + 4) {
            throw new CryptoOperationException("Decrypted content is malformed (missing filename header)");
        }
        ByteBuffer buf = ByteBuffer.wrap(packed);
        buf.get();
        int nameLen = buf.getInt();
        if (nameLen < 0 || nameLen > packed.length - 1 - 4) {
            throw new CryptoOperationException("Decrypted content is malformed (invalid filename header)");
        }
        byte[] nameBytes = new byte[nameLen];
        buf.get(nameBytes);
        String filename = nameLen > 0 ? new String(nameBytes, StandardCharsets.UTF_8) : null;
        byte[] content = new byte[buf.remaining()];
        buf.get(content);
        return new UnpackedContent(filename, content);
    }

    private record UnpackedContent(String filename, byte[] content) {}
}

File: GoCryptoService/internal/service/encryption_service.go
package service

import (
	"context"
	"encoding/base64"
	"encoding/binary"
	"errors"
	"time"

	gocrypto "gocryptoservice/internal/crypto"
	"gocryptoservice/internal/keystore"
	"gocryptoservice/internal/model"
	"gocryptoservice/internal/repository"

	"crypto/x509"
)

type EncryptResult struct {
	OperationID string
	Ciphertext  string
	Timestamp   time.Time
}

type DecryptResult struct {
	OperationID string
	Plaintext   string
	Filename    string
	Timestamp   time.Time
}

type EncryptionService struct {
	keystore *keystore.Manager
	audit    *AuditService
	details  *repository.EncryptionDetailRepository
}

func NewEncryptionService(ks *keystore.Manager, audit *AuditService, details *repository.EncryptionDetailRepository) *EncryptionService {
	return &EncryptionService{keystore: ks, audit: audit, details: details}
}

func (s *EncryptionService) Encrypt(ctx context.Context, plaintextB64, recipientCertB64, filename string) (EncryptResult, error) {
	return Run(ctx, s.audit, model.OpEncrypt, nil, nil, "ENCRYPT_FAILED",
		func(op *model.CryptoOperation) (EncryptResult, []byte, error) {
			plaintext, err := base64.StdEncoding.DecodeString(plaintextB64)
			if err != nil {
				return EncryptResult{}, nil, &ValidationError{Msg: "invalid base64 in plaintext: " + err.Error()}
			}
			certDER, err := base64.StdEncoding.DecodeString(recipientCertB64)
			if err != nil {
				return EncryptResult{}, nil, &ValidationError{Msg: "invalid base64 in recipientCertificate: " + err.Error()}
			}
			cert, err := x509.ParseCertificate(certDER)
			if err != nil {
				return EncryptResult{}, nil, NewCryptoOperationError("recipient certificate is invalid", err)
			}
			now := time.Now()
			if now.Before(cert.NotBefore) || now.After(cert.NotAfter) {
				return EncryptResult{}, nil, &CryptoOperationError{Msg: "recipient certificate is not valid: outside its validity period"}
			}

			packed := packWithFilename(plaintext, filename)
			ciphertext, err := gocrypto.Encrypt(packed, cert)
			if err != nil {
				return EncryptResult{}, nil, NewCryptoOperationError("encryption failed", err)
			}

			detail := &model.EncryptionDetail{
				OperationID: op.ID,
				Algorithm:   strPtr("AES256_GCM"),
				RecipientDN: strPtr(cert.Subject.String()),
				InputSize:   intPtr(len(plaintext)),
				OutputSize:  intPtr(len(ciphertext)),
			}
			if filename != "" {
				detail.OriginalFilename = strPtr(filename)
			}
			if err := s.details.Save(ctx, detail); err != nil {
				return EncryptResult{}, nil, err
			}

			result := EncryptResult{
				OperationID: op.ID,
				Ciphertext:  base64.StdEncoding.EncodeToString(ciphertext),
				Timestamp:   time.Now(),
			}
			return result, ciphertext, nil
		})
}

func (s *EncryptionService) Decrypt(ctx context.Context, ciphertextB64, keyAlias string) (DecryptResult, error) {
	alias := keyAlias
	return Run(ctx, s.audit, model.OpDecrypt, nil, &alias, "DECRYPT_FAILED",
		func(op *model.CryptoOperation) (DecryptResult, []byte, error) {
			ciphertext, err := base64.StdEncoding.DecodeString(ciphertextB64)
			if err != nil {
				return DecryptResult{}, nil, &ValidationError{Msg: "invalid base64 in ciphertext: " + err.Error()}
			}

			privKey, err := s.keystore.GetPrivateKey(keyAlias)
			if err != nil {
				if errors.Is(err, keystore.ErrKeyAliasNotFound) {
					return DecryptResult{}, nil, err
				}
				return DecryptResult{}, nil, NewKeystoreError("failed to load private key", err)
			}

			decrypted, err := gocrypto.Decrypt(ciphertext, privKey)
			if err != nil {
				return DecryptResult{}, nil, NewCryptoOperationError("decryption failed", err)
			}

			filename, content, err := unpackFilename(decrypted)
			if err != nil {
				return DecryptResult{}, nil, NewCryptoOperationError("decrypted content is malformed", err)
			}

			result := DecryptResult{
				OperationID: op.ID,
				Plaintext:   base64.StdEncoding.EncodeToString(content),
				Filename:    filename,
				Timestamp:   time.Now(),
			}
			return result, content, nil
		})
}

const filenamePackMarker byte = 0x01

func packWithFilename(content []byte, filename string) []byte {
	nameBytes := []byte(filename)
	packed := make([]byte, 0, 1+4+len(nameBytes)+len(content))
	packed = append(packed, filenamePackMarker)
	packed = binary.BigEndian.AppendUint32(packed, uint32(len(nameBytes)))
	packed = append(packed, nameBytes...)
	packed = append(packed, content...)
	return packed
}

func unpackFilename(packed []byte) (filename string, content []byte, err error) {
	if len(packed) == 0 || packed[0] != filenamePackMarker {
		return "", packed, nil
	}
	if len(packed) < 1+4 {
		return "", nil, errors.New("missing filename header")
	}
	nameLen := int(binary.BigEndian.Uint32(packed[1:5]))
	if nameLen < 0 || nameLen > len(packed)-1-4 {
		return "", nil, errors.New("invalid filename header")
	}
	name := packed[5 : 5+nameLen]
	rest := packed[5+nameLen:]
	return string(name), rest, nil
}

func strPtr(s string) *string { return &s }
func intPtr(i int) *int       { return &i }
